### ISSUE TITLE
Nested field support in visualizations

### ISSUE THREAD

**Issue by MerwaneHAMADI** [opensearch-project/OpenSearch-Dashboards#657]
**Is your feature request related to a problem? Please describe.**

As you know, nested support is not available in OpenSearch Dashboards Visualizations. [The community craves for it](https://github.com/elastic/kibana/issues/1084) because not everyone is working with unstructured data.

A lot of users work with a relation database as a source of truth they connect to Opensearch. So the data is relational by nature, and unless you can easily denormalize your index, nested objects are the recommended way to model this relationship along with joins (in some cases joins are better to easily update a relationship, or when the number of children for a given parent is really high).

So there is a strong need for nested field support in OpenSearch Dashboard in order to visualize this relational data.

**Describe the solution you'd like**

Being able to build nested queries and nested aggregations in the visualization dashboard.

**A clear and concise description of what you want to happen.**

We might want to learn from others that have tried to build the feature in the past, like ppadovani, filirom1 or SpaceManiac.
ppadovani suggested a [design](https://github.com/elastic/kibana/pull/5411#issuecomment-163305707) then implemented something like [this](https://ppadovani.github.io/knql_plugin/overview/) as a plugin along with a language called  [KNQL](https://ppadovani.github.io/knql_plugin/knql/).

This is a huge feature, so we could do it step by step instead of looking for the perfect solution : 
-  make ppadovani's plugin compatible with openSearch dashboard. Then identify the elements that make it hard to integrate it to the visualizations and try to find a fix for it.
- get feedback. To get even more feedback and make the plugin easily usable by all users, we could add a small checkbox(really hidden) on the settings to allow for it to be displayed in the visualizations, without having to do the steps to install the plugin. But users would need to be aware that it can break their dashboards if they use it.

I know it's not a really clear trajectory, but I can't really say where the feedbacks of the community are going to lead us to.
It looks like there are 3 options to tackle this problem. The interface can convert user interactions (clicks, drag and drop, typing) into :
1- an extended version of KQL. Currently, [KQL is currently only used to filter data](https://www.elastic.co/guide/en/kibana/current/kuery-query.html), and has no role in aggregating it
2- a separate language similar to [KNQL](https://ppadovani.github.io/knql_plugin/knql/). This language would be sent to the node web server which converts it back to a JSON sent to Elasticsearch.
3- json query directly, without using any language as a proxy.

Option 1 seems the ideal solution, because we don't create a separate language just for nested support. But the more pragmatic way of doing it might just be to have 2 different languages and THEN find a way to unify them into KQL.

**Describe alternatives you've considered**

I don't think there are alternatives to supporting nested fields natively on open search dashboard, but to cope waiting for it, (vega visualizations)[https://www.elastic.co/guide/en/kibana/current/vega.html] allow to build pretty much anything, if you are willing to spend the time and effort.

**Additional context**
This is the history of that feature request in Kibana : 
[Request in 2014]( https://github.com/elastic/kibana/issues/1084) :
[PR 1 by filirom1](https://github.com/elastic/kibana/pull/4645#issuecomment-132908544) : 
[PR 2 by ppadovani](https://github.com/elastic/kibana/pull/4806)
[PR 3 by ppadovani](https://github.com/elastic/kibana/pull/5411)
[Plugin] (https://ppadovani.github.io/knql_plugin/overview/) by ppadovani
This request is considered in Kibana as high hanging fruit (I agree) and **low impact** (I disagree). 

I will leave you with this screenshot which summarizes what you can get with the plugin right now :
![nested visualizations](https://user-images.githubusercontent.com/9652976/126495180-2ef1c10f-c64e-4f60-a503-6c1e692004d2.png)

**Comment by ahopp** [comment]
@MerwaneHAMADI I think the path you laid out (i.e. support plugin compatibility first as a proof of concept and iterate from there) seems like the correct approach. I'll tag as "help wanted". 

@ananzh I believe you mentioned that nested field support "Phase 1" was released in Kibana 7.6.0 but it's not enabled. Is that correct? Given we are forked 7.10.2, the assumption would be in the OpenSearch code but we need to make figure our what we'd need to do to enable it? Thoughts?

**Comment by MerwaneHAMADI** [comment]
@ahopp You are [correct](https://www.elastic.co/guide/en/kibana/7.x/release-notes-7.6.0.html).
It looks KQL supports nested queries right now.
If you add a simple nested object like this : 
```
PUT document/_doc/3
{
  "created_at": "2015-01-24",
  "object_content": [
    {
      "string_name": "document1",
      "string_type": "type1"
    },
    {
      "string_name": "document2",
      "string_type": "type2"
    }
  ]
}
```
We can see the KQL queries it properly : 
1 result as expected
![Screen Shot 2021-07-23 at 6 17 56 PM](https://user-images.githubusercontent.com/9652976/126853802-e2baaef6-3ee6-4407-8d2e-1c524ee5fae8.png)

0 result as expected
![Screen Shot 2021-07-23 at 6 17 37 PM](https://user-images.githubusercontent.com/9652976/126853803-6e43254d-c443-4e01-b94f-a7909332341e.png)

So that part is done. I realize I should precise what CAN be done and what can't be done right now.

I am coming back to you with a clear example. I think I am going to build the same example as in the KNQL (maybe add some items inside the room table,  so we get 2 levels of nesting). This was we can all use it as a reference.

![Screen Shot 2021-07-23 at 6 23 32 PM](https://user-images.githubusercontent.com/9652976/126853800-f70cc4b1-df0a-424b-bb97-d0fbdcf448d3.png)

**Comment by MerwaneHAMADI** [comment]
Also quick feedback for anyone starting to contribute : 
1- I thought of using docker for opensearch and git clone only dashboard, but the docker-compose.yml [here](https://hub.docker.com/r/opensearchproject/opensearch) has SSL implemented so unless you want to deal with authorization issues, you're better off just git clone opensearch as well. Normally, if you have the correct JAVA SDK, it should be straight forward.
@ahopp did you consider having a docker-compose.yml without SSL ? I tried to play with the configurations, but no way to disable SSL.
2- opensearch dashboard install should be straight forward. Just follow the instructions on github.
3- 
Questions : 
1- Anyone managed to connect intellij's debugger to opensearch dashboard ? It doesn't seem to work for some reason.
2- If you have any documentation or advice on how to start integrating the KNQL plugin into dashboard, I would really appreciate it. This is what the repo looks like : https://github.com/ppadovani/KibanaNestedSupportPlugin
3- When running opensearch with gradle and quitting gradle, my data is destroyed. I know this is not happening when using docker. Am I missing something ?

**Comment by ppadovani** [comment]
Hey folks this is my plug-in, and I'm happy to see some interest in trying to push this forward. 

Let me provide a list of what this plugin was capable of doing at the time I no longer had time to develop it.

- Full support for most visualizations and aggregations
- Knql provided a basic query language for building nested queries. There were limitations around more complex query types that were not supported like spans.

This plugin achieved this by modifying the index pattern stored by kibana to get the nested information into the internal config. It then overrode certain pieces of code in order to properly build out aggregations when a field referenced was nested.

One of the reasons I stopped developing this code was due to the shift in UI tech kibana was moving to. I'm not a UI dev, and lack the experience in what needed to be done to achieve the needed results. I also seem to recall that the kibana team began to limit what and how some pieces of code could be overridden which impeded my ability to inject the nested information that was needed.

I'd be more than happy to answer questions as best I can around what I did and how it worked.

**Comment by MerwaneHAMADI** [comment]
@ppadovani I am glad to see you here ! Thanks for your work on KNQL. I never got to use it, because our ES version is 7.7,  but I read all the github history and I think it's going in the right direction.
I have some questions :

1- What do you think is the most stable version of your plugin ? So we can test it ?

You wrote in your blog :
```
In general you will never notice that nested queries are being generated, as this is done for you in the query parser.
```
I like your syntax for that reason. Users don't need to know if it's nested or not. In my example above that's not the case : 

2- For phase 1 Do you think we should use the existing KQL syntax for now ?

3- How would you go about starting to put your plugin into opensearch dashboard and try and make it work ?

```
Edit: I now also recall that they decided to put the query parsing in elasticsearch itself rather than kibana, which made actually supporting aggregations impossible since kibana still didn't know what fields were nested.
```
4-  Are you saying that opensearch receives KQL and parses it there ? I thought OS dashboard was parsing KQL into JSON and sent it to opensearch.

**Comment by MerwaneHAMADI** [comment]
I just installed Elasticsearch and Kibana 6.4.3, ran bin/kibana-plugin install https://github.com/ppadovani/KibanaNestedSupportPlugin/releases/download/6.4.2-1.0.2/nested-fields-support-6.4.2-1.0.0.zip
And it looks like I can't enable the plugin.
![Screen Shot 2021-07-24 at 10 47 38 AM](https://user-images.githubusercontent.com/9652976/126877017-d3789052-a915-4e4f-823d-7f1462c26dfd.png)

I am on chrome, mac OS.
This guy had the same issue : 
https://github.com/ppadovani/KibanaNestedSupportPlugin/issues/106

So I am going to try version 5.6.11

**Comment by ppadovani** [comment]
> 
> 1- What do you think is the most stable version of your plugin ? So we can test it ? Edit : also a link to a docker-compose.yml of the elk installation would be great, to make installation easier.
> 

The latest released version was the most stable. As for a docker compose file, I'll have to go see if I have one kicking around. 
> 
> 
> 2- How would you go about starting to put your plugin into opensearch dashboard and try and make it work ?
> 
> 
I would start with the visualize and aggregation part before I worry about the query language as this provides the foundation that the query language can build on top of. This was from 3 years ago, so I'm dredging through my memories, but in general the following will need to be done in the base dashboard code:

- change, if not already done, the index pattern loading code to also pull the nested information from the mapping and store it in a consumable fashion. Most so this you should be able to steal right out of my plug-in code.
- the code for building the aggregations was centralized and used by most of the built in visualizations. Update this code to leverage the knowledge about nested fields from the first piece above. Again reference, unless the base code has changed a lot, the changes I injected into kibana.
- the UI for the visualizations only really needs help when dealing with a reverse nested agg. I had a way of doing this that should still be ok, but you should think about the usability of this a bit more.

Once you have this done, then I would look at the query language and decide if it's worth doing since kql does exist. Putting KNQL in will add an additional language the end users will have to know.(granted it's pretty simple syntax)

> 
> ```
> 
> Edit: I now also recall that they decided to put the query parsing in elasticsearch itself rather than kibana, which made actually supporting aggregations impossible since kibana still didn't know what fields were nested.
> 
> ```
> 
> 3-  Are you saying that opensearch receives KQL and parses it there ? I thought OS dashboard was parsing KQL into JSON and then sent it to opensearch.

It's been 3 years, so I may not have that part right. 

You know that you are forcing me to go back and look at the code and get involved again, right? 😉😂

**Comment by MerwaneHAMADI** [comment]
@ppadovani 
Haha ! Well I am not going to refuse some help, particularly because you probably know this problem better than me! But I also think you've already given a lot !

I realized it's just easier to download elasticsearch and kibana from the official website, without docker-compose.

As I said before, I got unlucky with the last release : version 6.4.2, so I am going to try the version just before that : 6.4.1

**Comment by ppadovani** [comment]
I can't look at this today due to family things. I should be able to pull my code and get it running again tomorrow morning.

**Comment by MerwaneHAMADI** [comment]
Thanks !
By the way I replaced this install command

`bin/kibana-plugin install https://github.com/ppadovani/KibanaNestedSupportPlugin/releases/download/6.4.2-1.0.2/nested-fields-support-6.4.2-1.0.0.zip
`
By this : 
`
kibana-6.4.2-darwin-x86_64/bin/kibana-plugin install https://github.com/ppadovani/KibanaNestedSupportPlugin/releases/download/6.4.2-1.0.2/nested-fields-support-6.4.2-1.0.2.zip
`

But still nothing. Looking at 6.4.1

**Comment by MerwaneHAMADI** [comment]
tldr;
- Opensearch dashboard doesn't even recognize include_in_root in visualizations anymore. Somehow it got removed between 6.4.2 and 7.10.2
- as @ppadovani said, the first step is to change the index pattern loading code to pull the nested information.

These are the simple steps to understand the feature I think opensearch dashboard should support : 
1- Install [elasticsearch 6.4.2](https://www.elastic.co/downloads/past-releases#elasticsearch) and [kibana 6.4.2](https://www.elastic.co/downloads/past-releases#kibana) => all you have to do is run bin/elasticsearch and bin/kibana.
2- Check that it's working normally => go to localhost:5601
3- install Ppadovani's plugin => /bin/kibana-plugin install https://github.com/ppadovani/KibanaNestedSupportPlugin/releases/download/6.4.2-1.0.2/nested-fields-support-6.4.2-1.0.2.zip
4- Follow the steps in this [video](https://www.youtube.com/watch?v=c0bOQgORmn8). 
5- Copy paste json requests [in this gist](https://gist.github.com/MerwaneHAMADI/21f7f93a67a6523e0b8c9aef994b2f98)
6- run them in dev tools
7- create an index pattern on houses
8- create a pie chart and slice by cars.make and then cars.options.name
I introduced an option called Acura Special. This option is only possible for Acura cars. When using include_in_root, it should show that this option is on all the car models, which is wrong.
9- Click on nested field support in index management
10- go back to the aggregation => TADA ! now it's showing that the option is only on the Acura model.
11- Now install opensearch and opensearch dashboard and run the seeds from the gist
12- Same thing, create an index pattern and create a pie chart. In this case, you CAN'T even see the include_in_root fields.

Everything is detailed in this video : 
https://www.youtube.com/watch?v=c0bOQgORmn8

**Comment by ppadovani** [comment]
Pulled the main of this repo, and started to look things over.. everything as been refactored from three years go, right?!? LOL

Anyway, the file `index_patter_field.ts` is the starting point that will need to be worked with. Go backward from there.. if I have time I might work on a branch.

**Comment by MerwaneHAMADI** [comment]
@ppadovani Don't hesitate to push even a PR where the plugin is setup and it does a simple hello world.
because I am not familiar with the way plugins interact with Opensearch dashboard or Kibana, so having a boilerplate is really valuable.

**Comment by ppadovani** [comment]
The PR is going to have to be against existing code... we need to have the nested path stored as part of the index pattern field. Once that is in place and functional, then the remaining code around aggregations can be modified to take advatage of it.

**Comment by MerwaneHAMADI** [comment]
Thanks for this information. I [first added the nested fields to the visualizations](https://github.com/MerwaneHAMADI/OpenSearch-Dashboards/pull/1/files), because it wasn't possible to see them.
The code is here : 
This is the result I have so far. As you can see, the acura special option is under all car brands, which shows that the aggregation is wrong right now.
By the way the only reason I have a result is because I set included_in_root to true for cars and cars.options.

Now I am working on agg_configs. It looks like most of the changes are inside the toDsl
I am trying to run kibana 6.4.2 dev version so I can compare with what's returned compared to osd.
I have a bazel issue though.

![Screen Shot 2021-07-31 at 2 35 28 PM](https://user-images.githubusercontent.com/9652976/127752947-d9d59476-9087-40d6-b141-db9dd2712973.png)

**Comment by MerwaneHAMADI** [comment]
It's not so obvious how to integrate the code to the toDSL. For now I am going to create from scratch just to understand a little better how the code works, but if you have a working version, don't hesitate to share it.

**Comment by MerwaneHAMADI** [comment]
So in this [branch](https://github.com/MerwaneHAMADI/OpenSearch-Dashboards/pull/1/commits/cee50caed40bc0d3a92a76984e3d8e32c2e8a192) I copied your agg_config and agg_configs files and converted them to typescript.

the value of dslTopLvl just before the return is : 
```
{
    "nested_2": {
        "nested": {
            "path": "cars"
        },
        "aggs": {
            "2": {
                "terms": {
                    "field": "cars.make",
                    "order": {
                        "_count": "desc"
                    },
                    "size": 5
                },
                "aggs": {
                    "nested_3": {
                        "nested": {
                            "path": "cars.options"
                        },
                        "aggs": {
                            "3": {
                                "terms": {
                                    "field": "cars.options.name",
                                    "order": {
                                        "_count": "desc"
                                    },
                                    "size": 10
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
```
So that looks ok, because this is exactly the aggregation I asked for.
But somehow elasticsearch fails to load response data. I think there is a missing piece.
![Screen Shot 2021-08-01 at 5 16 18 PM](https://user-images.githubusercontent.com/9652976/127789799-a816b96b-330a-4fb0-858e-1eafaafb6b29.png)

**Comment by ppadovani** [comment]
This looks good so far. Sorry I haven't been able to do much, life gets in the way. 

What you want to do is look at the agg_response code in my repo and compare against the existing implementation. This is the code that handles the response from
Elastic and formats it correctly. Pay attention to the two sub dirs.. tabify is what's used to create the table of data you can see on visualizations.

**Comment by MerwaneHAMADI** [comment]
Thanks ! No worries !
I have a question : What did you change in agg_response/tabify/_buckets.js ? It looks like it's the same.

**Comment by ppadovani** [comment]
I believe that was a clone in order to get around a weird access issue I had. So I cloned and owned it.

**Comment by MerwaneHAMADI** [comment]
ok thanks for the information. I am on build_hierarchical_data and it really looks different.

It looks like the most important code of build hierarchical data is here : 
const aggData = resp.aggregations[firstAgg.id] || resp.aggregations['nested_' + firstAgg.id][firstAgg.id];
const split = buildSplit(agg, metric, bucket[agg.id] || resp.aggregations['nested_' + firstAgg.id][firstAgg.id]);

but the buildSplit function disappeared...
At any point, you can see where I am at here : https://github.com/MerwaneHAMADI/OpenSearch-Dashboards/pull/1/commits

So don't hesitate if you see something incorrect

**Comment by ppadovani** [comment]
So kibana named the aggs based on the order they were defined as numbers. What I did was prefix nested_ to those levels that were, well, nested in ores to pull them back out on the response correctly. 

I'll try and grab some time first thing in the morning and look at your branch.

**Comment by MerwaneHAMADI** [comment]
I just realized what's wrong : the ui folder got removed : 
https://github.com/elastic/kibana/tree/6.8/src/ui => in 6.8 it was here 
in 7 it's gone : 
https://github.com/elastic/kibana/tree/7.0/src/ui

**Comment by ppadovani** [comment]
Looks like it got moved to here: [https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/plugins/data/common/search/aggs](https://github.com/opensearch-project/OpenSearch-Dashboards/tree/main/src/plugins/data/common/search/aggs)

All of the code that I overrode I used the same file names... so searching for date_histogram allowed me to locate the code that needs to be modified.

**Comment by MerwaneHAMADI** [comment]
```
const aggData = resp.aggregations[firstAgg.id] || resp.aggregations['nested_' + firstAgg.id][firstAgg.id];
const split = buildSplit(agg, metric, bucket[agg.id] || resp.aggregations['nested_' + firstAgg.id][firstAgg.id]);
```
For these 2 lines, it's not so obvious where to attach them in build_hierarchichal_data.ts :
- resp is not used as a variable name anymore, neither is aggData.
- I am looking for buildSplit accross the whole app, and it's only mentioned in the legacy plugins.

I am debugging build_hierarchichal_data to see where to stick prefix with '_nested'.

**Comment by MerwaneHAMADI** [comment]
@ppadovani I still didn't add build_hierarchichal_data, but it's already working.
I managed to get a chart, and I checked that this is correct.

So what is the use case that build_hierarchical_data enables ?

@ahopp is there some kind of documentation around building plugins on osd ?

**Comment by MerwaneHAMADI** [comment]
Are hacks still possible on opensearch dashboards ? Because it looks like that's what we need to build the plugin.

Is there some kind of boilerplate working on opensearch dashboard where a hack is implemented and where a method is overwritten ?

**Comment by MerwaneHAMADI** [comment]
to give more context, the plugin for nested support is defined like this :

```
return new kibana.Plugin({
    require: ['elasticsearch'],
    name: 'nested-fields-support',
    uiExports: {

      docViews: ['plugins/nested-fields-support/nested_support/doc_view/structure'],

      managementSections: [
        'plugins/nested-fields-support/index_pattern/management',
        'plugins/nested-fields-support/discover/management'
      ],

      hacks: [
        'plugins/nested-fields-support/nested_support'
      ]


    },

    config(Joi) {
      return Joi.object({
        enabled: Joi.boolean().default(true),
        index: Joi.string().default('.kibana')
      }).default();
    },


    // Update the .kibana index-pattern type to include a new nested flag
    init(server, options) {
      const { callWithInternalUser } = server.plugins.elasticsearch.getCluster('admin');

      updateIndexSchema(callWithInternalUser, server);

      server.route({
        path: '/api/nested-fields-support/mappings/{name}',
        method: 'GET',
        handler(req, reply) {
          const { callWithRequest } = server.plugins.elasticsearch.getCluster('admin');
          callWithRequest(req, 'indices.getMapping', {
            index: req.params.name
          }).then(function (response) {
            reply(response);
          });
        }
      });
    }

  });
```
But I can't find ui_exports on opensearch dashboard. has it been replaced by something else ?

**Comment by ppadovani** [comment]
You don't need to do any of this I believe. This code was all about turning on nested fields in an index pattern and loading the nested paths. This is already done in the base kibana code. We just need to make sure the visualizations/aggregations support detecting and handling nested fields. 

In terms of hacking via plugins, I think that is the wrong approach. The support for nested aggregations must be baked into the main code itself and not hacked in with a plugin. This aligns with the nested path support in the index pattern. 

Support for nested fileds in Discovery will take more effort and UI work as representing a nested structure, given the existing layout, is challenging.

**Comment by MerwaneHAMADI** [comment]
In terms of hacking via plugins, I think that is the wrong approach. The support for nested aggregations must be baked into the main code itself and not hacked in with a plugin. This aligns with the nested path support in the index pattern.



> You don't need to do any of this I believe. This code was all about turning on nested fields in an index pattern and loading the nested paths. This is already done in the base kibana code. We just need to make sure the visualizations/aggregations support detecting and handling nested fields.

Ok, good to know. 

> In terms of hacking via plugins, I think that is the wrong approach. The support for nested aggregations must be baked into the main code itself and not hacked in with a plugin. This aligns with the nested path support in the index pattern.

After reading all the issues you had merging this feature into kibana, I was thinking building a plugin was the correct way. But now, I realize most of the work has already been done : the index pattern already displays all we need. So I agree with you : implementing a plugin that hacks osd is not necessary anymore.


So now the question is what is the minimum viable implementation that aligns with the vision of opensearch dashboard and allows us to get momentum on this issue ?
Option 1 
- We don't merge anything into the code until all visualizations are supported. It might take more time to have a really clean implementation. I have looked a bit at all the visualizations, and I think there are some edge cases that would require some time to explore, before saying they are officially supported. I think this option might be too ambitious
Option 2 
- We choose one visualization (pie chart for example), and we support it fully. This means in the pie chart visualization, users will be able to choose nested fields, but not in the table visualization for example. Then we add visualizations one by one.

@ahopp @stockholmux @stockholmux what do you think ? 

I am going to drill down into option 2. If we go for it, what work is remaining : 
This [branch](https://github.com/MerwaneHAMADI/OpenSearch-Dashboards/pull/1/files) is working but : 
- nested fields should only be displayed for pie charts.
- when you drag and drop slices of a pie chart, this should give a correct result  => cf this [comment](https://github.com/elastic/kibana/pull/5411#issuecomment-162944976)
- when you click on the pie chart, it should filter correctly. => the current behavior on my branch is that it's going to try and filter the element in the root of the document. This means this will only work if you include_in_root your nested objects. Otherwise it will display no results, which is pretty confusing.
For the last item we have 2 options : 
Option 1 
When the user clicks on a pie chart, it creates a nested filter.
Option 2
We know the vision is option 1, but in the meantime, we don't allow the user from filtering on the pie chart. TO be clear, we only prevent the user from filtering if he's filtering on a nested field. If the field is "normal", then it should behave as before, otherwise this would be considered a regression.
Option 3 : 
Same as option 2, but we only prevent the user from filtering on the pie chart IF the field is not included in root (maybe too complex to implement)
Option 4 (in an ideal world)
Clicking on the pie chart not only creates a nested filter, but also modifies the pie chart aggregation itself and adds a filter. Because a nested filter still wont display the pie chart correctly. when the user clicks on a pie, he wants this pie to become the new whole of the pie chart. => probably too hard for now.

I think it's between option 1 and 2

**Comment by stockholmux** [comment]
@MerwaneHAMADI I'd like to hear more voices here, but my gut says that iterative is the best approach - so, support the pie chart first. Option 2 on the second question makes sense - if you are running `include_in_root` you are progressively enhanced, if not you have the same thing - right?

**Comment by MerwaneHAMADI** [comment]
> if you are running include_in_root you are progressively enhanced.
Not sure what I made myself clear, so I a going to detail.

if we go for option 2 of not allowing nested pie charts to be clickable, then even if you have your object include_in_root, you won't be able to filter against it, because it's just not clickable.
But it's a strict improvement, in the sense that right now nobody can create nested pie charts. So if we give them non filterable nested pie charts, it's still an improvement, over no nested pie charts at all.
My concerns are the following => 
- if some part of the pie chart is non nested and some other is nested, we need to prevent the nested part from being filterable but the other part to still be filterable.


@ppadovani I want to hear what you think on that one. Do we try and create one PR to try and merge every visualization into the code or just start with the pie chart ?

If we only implement pie charts, there are still a lot of things to define on how to communicate this to users, and if we prevent the users from filtering on nested pie charts (and I insist : only nested pie charts, not regular pie charts, otherwise it's a regression), then they need to understand why.
This is why I am still doubting between option 1 and 2. It might be easier to also build a nested filter when the user clicks, instead of creating a workaround for it.

**Comment by MerwaneHAMADI** [comment]
Forget what I said above about nested filters on pie chart visualizations : it's actually working. I tested it on fields that are not included in root and it works.
This means when someone clicks on a pie chart, it actually selects only the records that match with the criteria in the pie charts, and it's automatically using the filter of the pie chart. Obviously, it's still not smart enough to select the pie chart as a whole, but I don't think this is in the scope of what we're trying to do (cf option 4 above).

This is a pretty good news. I am still testing it on all the fields.

**Comment by ahopp** [comment]
@MerwaneHAMADI RE: viable implementation, in general my bias is "progress not perfection"! I think the incremental approach would be my preference. It seems to be the best path for feedback and iteration - and learn what works well as what we'd want to change before expanding to all visualizations. 

RE: Communicating to users, did you have a preferences? One treatment could be some tag with qualifiers on the CX treatment (ex. a beta tooltip) as we expand the feature. We would also include the changes in the release notes but could qualify the feature scope. There is some nuance to add here but one option. 

This is exciting stuff - thanks for driving!

**Comment by ppadovani** [comment]
Just to be clear... the field support for nested is already baked in. Adding the support to the visualization controls for buckets/metrics and the updates to the agg_config.ts code will enable nested aggregations for more than the pie chart, as I believe the code works this way. So I don't think you can just turn on support for pie charts. 

Personally I would add the basic support, and run through some testing to make sure it doesn't break existing aggs/charts/etc. Then I would start collecting issues/features that are broken/missing in the context of nested and start burning that list down.

**Comment by MerwaneHAMADI** [comment]
@ahopp ok, so you also lean towards starting with the pie chart.
And I agree with what you said for the UX treatment. Fortunately, the biggest UX hurdle is gone, now that the user can actually filter on nested pie charts.

@ppadovani There is a way to not display nested fields for specific visualizations. in the field.ts react component, you can allow nested fields based on the query parameter ?type=pie (for example). So it's possible to just turn on support just for pie charts, and even only for a specific aggregations.

> Personally I would add the basic support, and run through some testing to make sure it doesn't break existing aggs/charts/etc. Then I would start collecting issues/features that are broken/missing in the context of nested and start burning that list down.

Do you mean collecting issues/features from people using it in the master branch ? I am a bit worried about that. I am not sure of the impact of that feature on the user experience. For now I am still not able to find any regression. I can't even reproduce [this](https://github.com/elastic/kibana/pull/5411#issuecomment-162944976).

But I don't know what can go wrong with this feature.

This might sound overly cautious, but what I am the most comfortable doing right now is shipping **nested pie chart with histograms only** . This would allow to have a proof of concept, make real users use it.
And histograms with pie charts are not used as much as terms aggregations for example. So if something breaks (which I will make sure it doesn't, because I am testing it extensively), at least we only break one aggregation of a single visualization.
Then if we don't find any regression, we can unlock the full pie chart at once.

The logic I suggest we implement is :
- if the user is inside the pie chart and the user selected histograms, show him all the nested fields. Otherwise, only show him the regular fields.
- I don't think adding this logic will add significantly more latency.

Again, it could just be a matter of weeks to then start adding even more aggregations and visualizations, believe me, I am the first to need it.

**Comment by MerwaneHAMADI** [comment]
This bug happens in the main branch : 
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/718

It's good to know it now, because we'll know it's not a regression from the nested field support.

**Comment by ahopp** [comment]
@MerwaneHAMADI Seems like you're making great progress here! Anything we can do to help / support?

**Comment by nhirshler** [comment]
What is the current status of nested fields in ELK?
I installed elasticsearch and kibana 7.16.2 and I cannot access nested fields in visualizatation. It worked in the older version of elastcsearch and kibana but after this last update is not longer possible to create term aggregation based on nested field.

**Comment by Jorgen-VikingGod** [comment]
Any update or plan to support nested fields in visualizations?

**Comment by heygambo** [comment]
I also really need this

**Comment by Mephilius** [comment]
Hi,

I also really need this urgend. We are a large corporation that plans to use elastic. The visualization of the nested fields in Kibana is a key point that we need.

**Comment by fatihakafou** [comment]
Hi,

This is also really needed for us. Any update on it ??

**Comment by zidludwig** [comment]
i was searching for long time how to put my nested objects with logstash into the DB to analyse it with kibana. Now i found the solution: "DON'T DO IT"! i will try to do with workarounds (split events,..) but i'd really love to it straight forward !

**Comment by kavilla** [comment]
[Triage]

@seanneumann @ahopp, as we can see this a high ask for the community. Could we get this prioritized?

**Comment by JuanAntonio03** [comment]
Hello,

this is something we desperately need in my team. ¿¿ Any update ??

**Comment by mik3fly** [comment]
Hello, is there any update or roadmap to implement this ? thanks

**Comment by b-akhil** [comment]
Hello. Is there any update regarding the support for aggregations on nested fields within OpenSearch Dashboards ?

**Comment by AdaptiveStep** [comment]
Hi! Does anyone know any good workaround to this at least? I'm thinking of splitting the events (denormalizing), but i'd much rather avoid it. Any clues?

**Comment by Fantus** [comment]
Same here, we would love to have this feature

**Comment by joaopfcruz** [comment]
+1 here.

Any updates on this?

**Comment by RawanS** [comment]
Same here, please let us know of any updates on this or if there is some workaround. Thanks.

**Comment by suraj-mathew** [comment]
It has been 2.5 years that everyone is asking for a much needed enhancement. But sadly no direct answer so far. 
It will be good if any of the following replies can be made 

- Yes, it will be implemented
- No, it will not be implemented
- Still working on for a solution

so that people will get a notification whether to wait or go-ahead with some other solution (if any).

**Comment by ahopp** [comment]
I think it's safe to say that @merwanehamadi is no longer working on this issue - we can all see there hasn't been much activity on it recently and there are quite a few unresolved inquires about its current status. It's also flagged as 'help wanted' which typically means there are no current assignees working on the dev side. In the meantime, we might want to mark it as part of issues tagged 'backlog'? Thoughts - @joshuarrrr @ananzh @kavilla @ashwin-pc

**Comment by ahopp** [comment]
Also, if anyone is interested in helping out with this issue or want to contribute to the resolution of this issue, please feel free. Everyone in the community is welcome to make contributions. And I don't think @seanneumann is the right assignee - I don't thinks he's driving from the dev effort on the AWS side.

**Comment by nkalpakis21** [comment]
i see this has been open for awhile. is there trouble getting a dev to fix this issue? what is blocking this for so long

**Comment by mmguero** [comment]
It's 2026, is there any on movement on this? This really limits usefulness of the nested field type and nested aggregations.

**Comment by xluo-aws** [comment]
Will take a look next week when our tech lead is back from vacation

**Comment by ruanyl** [comment]
We’re build a new visualization framework(data-driven, query-agnostic) which we could add support to visualize nested data with Sunburst or Treemap type of charts, however, these chart types are not prioritized in current roadmap. We will check with stakeholders in terms prioritization.
